packages/typescript/internal/ts_project.bzl

```diff
@@ -756,7 +783,29 @@ def ts_project_macro(
     if declaration_map:
         typing_maps_outs.extend(_out_paths(srcs, typings_out_dir, root_dir, allow_js, {"*": ".d.ts.map"}))
 
-    if not len(js_outs) and not len(typings_outs):
+    if type(transpiler) == "function" or type(transpiler) == "rule":
+        tsc_js_outs = []
+        tsc_map_outs = []
+        transpiler(
+            name = name + "_transpile",
+            srcs = srcs,
+            js_outs = js_outs,
+            map_outs = map_outs,
+            data = [],
+            tags = kwargs.get("tags", []),
+            visibility = kwargs.get("visibility", None),
+            testonly = kwargs.get("testonly", None),
+            **transpiler_kwargs
```

Comment:
This is getting into the weeds but if you wanted to pass different `tags` to the transpiler then you could in theory write
```
ts_project(
   name = "foo",
   tags = ["foo"],
   transpiler_kwargs = {
      tags: ["bar"],
   }
```
and this could be supported if `kwargs.get("tags", [])` was merged into `transpiler_kwargs` if and only if `transpiler_kwargs` does not already have a `tags`.

Reply:
Complete list of "common" attributes to automatically merge into `transpiler_kwargs` is https://docs.bazel.build/versions/main/be/common-definitions.html#common-attributes

Reply:
once we are passing through so much to the transpiler, it starts feeling like this macro is too leaky, and the user should just write their pair of typecheck(emitDeclarationOnly=true) and transpile() rules in their BUILD file.

Reply:
How about instead of hiding this complexity inside ts_project, we expect users to write a simple wrapper and pass it to the transpiler?  in the case of `.swcrc` i believe a macro like below can be written once and reused everywhere. 

```python
def swc_with_swc_rc(**kwargs):
     return swc(swcrc = "//.swcrc", **kwargs)
```

Reply:
@tjgq I wonder if you have thoughts about this. I think we discussed the idea once?

Reply:
oh @thesayyn you can use https://github.com/bazelbuild/bazel-skylib/blob/main/lib/partial.bzl for this, so that the swcrc argument is bound where you make the higher-order function, then ts_project calls that partial.

(Bazel 5 will be nicer here since you could just use a lambda)

PTAL